### ISSUE TITLE
Use Public Suffix List to Calculate Domain

### ISSUE THREAD

**Issue by alexyao2015** [caddyserver/certmagic#144]
## What would you like to have changed?

Instead of using dns to obtain the primary zone, use the public suffix list [here](https://publicsuffix.org/)  to obtain the domain. This is the same way that Let's Encrypt calculates the registered domain discussed [here](https://letsencrypt.org/docs/rate-limits/) .

Keeping the current method of obtaining the primary zone using DNS could still be useful if obtaining this list is blocked by a firewall for some reason.

## Why is this feature a useful, necessary, and/or important addition to this project?

This alleviates some issues with split brain DNS which could result in incorrect resolution and also removes the need to manually specify DNS servers for domain resolution when used. Also discussed at caddy-dns/cloudflare#12

**Comment by francislavoie** [comment]
We're not using PSL, on purpose: https://github.com/sleevi/psl-problems

**Comment by alexyao2015** [comment]
From that article, is there a reason why the RZD is not used instead? For the purposes of Let's Encrypt, the PSL does seem compelling, simply because it is used by Let's Encrypt and isn't used in a security or usability context as mentioned in the article. The uses of it here don't seem to apply.

**Comment by mholt** [comment]
Hmmmm. Thanks for the suggestion, but my inclination is to decline this request. Given the drawbacks of the PSL, I'm not sure configuring the PSL is going to be any improvement over simply configuring the DNS resolvers in the edge case of split-brain DNS. Both require proper network/firewall/DNS configuration nonetheless, and it just adds more complexity that is still brittle.

**Comment by alexyao2015** [comment]
It could be embedded directly in the program, removing network access requirements.

**Comment by mholt** [comment]
No, the PSL changes, so it would need to be fetched from an authoritative, current source.
